### DISCUSSION TITLE
Anvil inventories still don't work like they could

### DISCUSSION THREAD

**Discussion by NonSwag** [PaperMC/Paper#10600] [Ideas]
### Is your feature request related to a problem?

So, there's been an ongoing issue with Bukkit and anvil inventories that seems to have persisted for quite some time. It's been rather frustrating, to say the least. Essentially, when attempting to create an anvil inventory using Bukkit's createInventory method, things just don't function as expected.

The crux of the problem lies in anvils created this way ending up as type CraftInventoryCustom instead of the expected CraftInventoryAnvil. This leads to a myriad of issues – events fail to trigger properly, and window properties appear to be behaving strangely. It's all quite bothersome.

Currently, the most dependable method to obtain a functional anvil is by utilizing Player#openAnvil. However, let's be honest, that's not a universal solution. There are instances where greater flexibility is required.

### Describe the solution you'd like.

It would be incredibly helpful to have a method like Bukkit#createAnvil or perhaps a solution to address the issues in the existing createInventory methods.

### Describe alternatives you've considered.

Currently, the way to go is by using NMS, which seems a bit silly for such a simple task.

**Comment by Sytm** [comment]
> However, let's be honest, that's not a universal solution. There are instances where greater flexibility is required.

Could you perhaps expand on the problems you have with this method?
I use that method to open a text input GUI and have no problems customizing the behavior with the different inventory events that are already in the Paper API

**Comment by NonSwag** [comment]
in some cases you want reusable inventories or custom inventory holders
also it is not possible to define a custom name for the inv when using player open anvil

**Comment by Sytm** [comment]
Yeah, applying a custom name is currently very annoying, because you need to override the title in the InventoryOpenEvent, but I already opened a PR that alleviates this problem (#9658).

~~Iirc making the inventory reusable is not really possible, because the real inventory is tied to a player~~, but custom inventory holders would be indeed nice for easier handling of events

**Comment by NonSwag** [comment]
I think anvils are not player but location bound

**Comment by rayanbzd** [comment]
This is sadly still an issue. Check my comment here: https://github.com/PaperMC/Paper/issues/5284#issuecomment-2113036334

**Comment by Machine-Maker** [comment]
Going to close this for now. Upstream has merged in the [MenuType API](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/commits/45c206f2c56c8e38130ba2e0af593759c9864fc0#src%2Fmain%2Fjava%2Forg%2Fbukkit%2Fcraftbukkit%2Finventory%2Futil%2FCraftMenuBuilder.java) which should solve this, create AnvilView's via that API (once it's upstream'd into paper).
